### PR TITLE
NOISSUE - Implement Thing Delete

### PR DESCRIPTION
**Pull request by arvindh123**
# What type of PR is this?
Implements thing delete

## Have you included tests for your changes?

Yes, I have included tests for my changes.

## Did you document any new/modified feature?

No, Not Planned in this PR.

### CODE REVIEW

File: pkg/sdk/go/sdk.go [hunk 1]
```diff
@@ -453,6 +453,13 @@ type SDK interface {
 	//  fmt.Println(users)
 	ListThingUsers(thingID string, pm PageMetadata, token string) (UsersPage, errors.SDKError)
 
+	// DeleteThing delete given thing id.
```
Comment (ianmuchyri) [inline pkg/sdk/go/sdk.go:456]:
```suggestion
	// DeleteThing deletes a thing with the given id.
```

File: things/api/http/responses.go [hunk 2]
```diff
@@ -138,20 +138,34 @@ func (res viewMembersRes) Empty() bool {
 	return false
 }
 
-type deleteClientRes struct {
+type changeStatusClientRes struct {
```
Comment (ianmuchyri) [inline things/api/http/responses.go:141]:
```suggestion
type changeClientStatusRes struct {
```

File: things/postgres/clients.go [hunk 3]
```diff
@@ -30,6 +30,9 @@ type Repository interface {
 
 	// RetrieveBySecret retrieves a client based on the secret (key).
 	RetrieveBySecret(ctx context.Context, key string) (mgclients.Client, error)
+
+	// Delete deletes given client id
```
Comment (ianmuchyri) [inline things/postgres/clients.go:34]:
```suggestion
	// Delete deletes client with given id
```

File: things/things.go [hunk 4]
```diff
@@ -57,6 +57,9 @@ type Service interface {
 
 	// Authorize used for AuthZ gRPC server implementation and Things authorization.
 	Authorize(ctx context.Context, req *magistrala.AuthorizeReq) (string, error)
+
+	// DeleteClient delete given client ID.
```
Comment (ianmuchyri) [inline things/things.go:61]:
```suggestion
	// DeleteClient deletes client with given ID.
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/sdk.go (hunk 1 above)
Comment (arvindh123) [reply pkg/sdk/go/sdk.go]: done

File: things/api/http/responses.go (hunk 2 above)
Comment (arvindh123) [reply things/api/http/responses.go]: done

File: things/postgres/clients.go (hunk 3 above)
Comment (arvindh123) [reply things/postgres/clients.go]: done

File: things/things.go (hunk 4 above)
Comment (arvindh123) [reply things/things.go]: done

File: api/openapi/things.yml [hunk 5]
```diff
@@ -169,7 +169,26 @@ paths:
           description: Missing or invalid content type.
         "500":
           $ref: "#/components/responses/ServiceError"
-
+    delete:
+      summary: Delete thing for given thing id.
```
Comment (ianmuchyri) [inline api/openapi/things.yml:173]:
```suggestion
      summary: Delete thing for a thing with the given id.
```

File: api/openapi/things.yml [hunk 6]
```diff
@@ -169,7 +169,26 @@ paths:
...
+      description: |
+        Delete thing remove given thing id from repo
```
Comment (ianmuchyri) [inline api/openapi/things.yml:175]:
```suggestion
        Delete thing removes a thing with the given id from repo
```

File: api/openapi/things.yml [hunk 7]
```diff
@@ -169,7 +169,26 @@ paths:
...
+        and removes all the policies related to thing.
```
Comment (ianmuchyri) [inline api/openapi/things.yml:176]:
```suggestion
        and removes all the policies related to this thing.
```

File: things/api/http/responses.go [hunk 8]
```diff
@@ -138,20 +138,34 @@ func (res viewMembersRes) Empty() bool {
...
 
+func (res changeClientStatusRes) Headers() map[string]string {
+	return map[string]string{}
+}
+
+func (res changeClientStatusRes) Empty() bool {
+	return false
+}
+
+type deleteClientRes struct{}
```
Comment (rodneyosodo) [inline things/api/http/responses.go:157]: Add type check

File: things/events/events.go [hunk 9]
```diff
@@ -380,3 +381,14 @@ func (sce shareClientEvent) Encode() (map[string]interface{}, error) {
 		"user_ids":  strings.Join(sce.userIDs, ","),
 	}, nil
 }
+
+type removeClientEvent struct {
```
Comment (rodneyosodo) [inline things/events/events.go:385]: also add type check

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: things/service.go [hunk 10]
```diff
@@ -427,6 +427,55 @@ func (svc service) Unshare(ctx context.Context, token, id, relation string, user
...
+	if _, err := svc.auth.DeletePolicy(ctx, &magistrala.DeletePolicyReq{
+		SubjectType: auth.DomainType,
+		Object:      id,
+		ObjectType:  auth.ThingType,
+	}); err != nil {
+		return err
+	}
+
+	// Remove from cache
+	if err := svc.clientCache.Remove(ctx, id); err != nil {
```
Comment (rodneyosodo) [inline things/service.go:458]: We can invalidate the cache i.e (below authorize check) after auth check on L438

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: things/service.go (hunk 10 above)
Comment (arvindh123) [reply things/service.go]: done

File: things/events/events.go (hunk 9 above)
Comment (arvindh123) [reply things/events/events.go:386]: done

File: things/api/http/responses.go (hunk 8 above)
Comment (arvindh123) [reply things/api/http/responses.go:158]: done

File: api/openapi/things.yml [hunk 11]
```diff
@@ -169,7 +169,26 @@ paths:
...
+        and removes all the policies related to this thing.
+      tags:
+        - Things
+      parameters:
+        - $ref: "#/components/parameters/ThingID"
+      security:
+        - bearerAuth: []
+      responses:
+        "204":
+          $ref: "#/components/responses/ThingDeleteRes"
```
Comment (rodneyosodo) [inline api/openapi/things.yml:185]: No need to reference this since it it only description
```suggestion
          description: Thing deleted.
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: api/openapi/things.yml (hunk 11 above)
Comment (arvindh123) [reply api/openapi/things.yml]: done

**Review by rodneyosodo — APPROVED** [review]

**Review by ianmuchyri — APPROVED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**